pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -140,17 +143,101 @@ func readConfig(config io.Reader) (Config, error) {
 	return cfg, err
 }
 
+// parseMetadataUUID reads JSON from OpenStack metadata server and parses
+// instance ID out of it.
+func parseMetadataUUID(jsonData []byte) (string, error) {
+	// We should receive an object with { 'uuid': '<uuid> } and couple of other
+	// properties (which we ignore).
+
+	var obj map[string]interface{}
+	err := json.Unmarshal(jsonData, &obj)
+	if err != nil {
+		return "", err
+	}
+
+	tmp, ok := obj["uuid"]
+	if !ok {
+		err = fmt.Errorf("cannot parse OpenStack metadata, no uuid found")
+		return "", err
+	}
+
+	uuid, ok := tmp.(string)
+	if !ok {
+		err = fmt.Errorf("cannot parse OpenStack metadata, uuid has invalid format")
+		return "", err
+	}
+
+	if uuid == "" {
+		err = fmt.Errorf("got empty uuid")
+		return "", err
+	}
+
+	return uuid, nil
+}
+
+func readInstanceID() (string, error) {
+	// Try to find instance ID on the local filesystem (created by cloud-init)
+	const instanceIDFile = "/var/lib/cloud/data/instance-id"
+	idBytes, err := ioutil.ReadFile(instanceIDFile)
+	if err == nil {
+		instanceID := string(idBytes)
+		instanceID = strings.TrimSpace(instanceID)
+		glog.V(3).Infof("Got instance id from %s: %s", instanceIDFile, instanceID)
+		if instanceID != "" {
+			return instanceID, nil
+		}
+		// Fall through with empty instanceID and try metadata server.
+	}
+	glog.V(5).Infof("Cannot read %s: '%v', trying metadata server", instanceIDFile, err)
+
+	// Try to get JSON from metdata server.
+	const metadataUrl = "http://169.254.169.254/openstack/2012-08-10/meta_data.json"
```

Comment:
Yes, that's the only IP address where metadata service can be found. http://docs.openstack.org/admin-guide-cloud/compute-networking-nova.html#metadata-service